File: backend/app/models/patient.py
import enum
import uuid
from datetime import date, datetime
from sqlalchemy import Date, DateTime, Enum, String, Boolean
from sqlalchemy.orm import Mapped, mapped_column

from app.config.database import Base


class PatientSex(str, enum.Enum):
    M = "M"
    F = "F"


class Patient(Base):
    __tablename__ = "patients"

    id: Mapped[str] = mapped_column(String(36), primary_key=True, default=lambda: str(uuid.uuid4()))
    record_number: Mapped[str] = mapped_column(String(50), unique=True, nullable=False, index=True)
    first_name: Mapped[str] = mapped_column(String(100), nullable=False)
    last_name: Mapped[str] = mapped_column(String(100), nullable=False)
    birth_date: Mapped[date] = mapped_column(Date, nullable=False)
    sex: Mapped[PatientSex] = mapped_column(Enum(PatientSex), nullable=False)
    email: Mapped[str] = mapped_column(String(255), nullable=False)
    phone: Mapped[str] = mapped_column(String(30), nullable=False)
    city: Mapped[str] = mapped_column(String(120), nullable=False)
    address: Mapped[str] = mapped_column(String(255), nullable=True)
    insurance_number: Mapped[str | None] = mapped_column(String(50), nullable=True)
    is_active: Mapped[bool] = mapped_column(Boolean, default=True, nullable=False)
    created_at: Mapped[datetime] = mapped_column(DateTime, default=datetime.utcnow, nullable=False)
    updated_at: Mapped[datetime] = mapped_column(
        DateTime, default=datetime.utcnow, onupdate=datetime.utcnow, nullable=False
    )

File: frontend/services/api/patient.ts
import apiClient from './client';

export interface Patient {
  id: string;
  record_number: string;
  first_name: string;
  last_name: string;
  birth_date: string;
  sex: string;
  email: string;
  phone: string;
  city: string;
  address?: string;
  insurance_number?: string;
  is_active: boolean;
  created_at: string;
  updated_at: string;
}

export interface PatientsResponse {
  items: Patient[];
  total: number;
  page: number;
  limit: number;
  pages: number;
}

const patientService = {
  async getPatients(page: number = 1, limit: number = 10, search: string = ''): Promise<PatientsResponse> {
    const response = await apiClient.get<PatientsResponse>('/patients', {
      params: { page, limit, search }
    });
    return response.data;
  },

  async getPatient(id: string): Promise<Patient> {
    const response = await apiClient.get<Patient>(`/patients/${id}`);
    return response.data;
  },

  async createPatient(data: any): Promise<Patient> {
    const response = await apiClient.post<Patient>('/patients', data);
    return response.data;
  },

  async updatePatient(id: string, data: Partial<Patient>): Promise<Patient> {
    const response = await apiClient.put<Patient>(`/patients/${id}`, data);
    return response.data;
  },

  async deletePatient(id: string): Promise<{ message: string }> {
    const response = await apiClient.delete<{ message: string }>(`/patients/${id}`);
    return response.data;
  }
};

export default patientService;
